src/main/java/eu/europeana/metis/sandbox/service/workflow/EnrichmentServiceImpl.java

```diff
@@ -3,41 +3,64 @@
 import static java.util.Objects.requireNonNull;
 
 import eu.europeana.enrichment.rest.client.EnrichmentWorker;
-import eu.europeana.enrichment.rest.client.exceptions.DereferenceException;
-import eu.europeana.enrichment.rest.client.exceptions.EnrichmentException;
+import eu.europeana.enrichment.rest.client.report.ProcessedResult;
+import eu.europeana.enrichment.rest.client.report.Report;
+import eu.europeana.enrichment.rest.client.report.Type;
 import eu.europeana.metis.sandbox.common.exception.RecordProcessingException;
+import eu.europeana.metis.sandbox.common.exception.ServiceException;
 import eu.europeana.metis.sandbox.domain.Record;
 import eu.europeana.metis.sandbox.domain.RecordError;
 import eu.europeana.metis.sandbox.domain.RecordInfo;
-import eu.europeana.metis.schema.convert.SerializationException;
 import java.util.LinkedList;
 import java.util.List;
+import java.util.Optional;
+import java.util.Set;
+import java.util.Objects;
+
 import org.springframework.stereotype.Service;
 
 @Service
 class EnrichmentServiceImpl implements EnrichmentService {
 
-  private final EnrichmentWorker enrichmentWorker;
-
-  public EnrichmentServiceImpl(EnrichmentWorker enrichmentWorker) {
-    this.enrichmentWorker = enrichmentWorker;
-  }
-
-  @Override
-  public RecordInfo enrich(Record record) {
-    requireNonNull(record, "Record must not be null");
-
-    List<RecordError> recordErrors = new LinkedList<>();
-    byte[] result;
-    try {
-      result = enrichmentWorker.process(record.getContentInputStream());
-    } catch (EnrichmentException|DereferenceException|SerializationException e) {
-      result = record.getContent();
-      recordErrors.add(new RecordError(new RecordProcessingException(record.getProviderId(), e)));
-    } catch (Exception e) {
-      throw new RecordProcessingException(record.getProviderId(), e);
+    private final EnrichmentWorker enrichmentWorker;
+
+    public EnrichmentServiceImpl(EnrichmentWorker enrichmentWorker) {
+        this.enrichmentWorker = enrichmentWorker;
+    }
+
+    @Override
+    public RecordInfo enrich(Record record) {
+        requireNonNull(record, "Record must not be null");
+
+        List<RecordError> recordErrors = new LinkedList<>();
+        ProcessedResult<byte[]> result;
+        try {
+            result = enrichmentWorker.process(record.getContentInputStream());
+            Set<Report> reports = result.getReport();
+
+            if (result.getRecordStatus().equals(ProcessedResult.RecordStatus.STOP)) {
+                Optional<Report> report = reports.stream().filter(rep -> Objects.equals(rep.getMessageType(), Type.ERROR)).findFirst();
+                throw new RecordProcessingException(record.getProviderId(), handleRecordStopException(report));
+            }
```

Comment:
Wouldn't it be nicer to move all this code, including the if statement, to the `handleRecordStopException` method below? Then you don't have to pass an optional (but just the reports) and you don't need a return value (the method optionally throws an exception).

Reply:
That's a good idea, I'll do that